modules/core/04-channel/keeper/upgrade.go

```diff
@@ -68,3 +73,187 @@ func (k Keeper) constructProposedUpgrade(ctx sdk.Context, portID, channelID stri
+	}
+
+	expectedChannel := types.Channel{
+		State:        types.INITUPGRADE,
+		Counterparty: types.NewCounterparty(channel.Counterparty.PortId, channel.Counterparty.ChannelId),
+		Ordering:     channel.Ordering,
+		// q: are we supposed to use the counterparty connection hops here? if so how to grab?
+		ConnectionHops:  counterpartyHops,
+		Version:         channel.Version,
+		UpgradeSequence: counterpartyUpgradeSequence,
+	}
+
+	// verify that the counterparty channel has correctly entered into the upgrade process
+	if err := k.connectionKeeper.VerifyChannelState(
+		ctx,
+		connectionEnd,
+		proofHeight,
+		proofCounterpartyChannel,
+		channel.Counterparty.PortId,
+		channel.Counterparty.ChannelId,
+		expectedChannel,
+	); err != nil {
+		return types.Upgrade{}, types.Channel{}, err
+	}
+
+	// verifies the proof that a particular proposed upgrade has been stored in the upgrade path.
+	if err := k.connectionKeeper.VerifyChannelUpgrade(ctx, connectionEnd, proofHeight, proofUpgrade, channel.Counterparty.PortId,
+		channel.Counterparty.ChannelId, counterpartyProposedUpgrade); err != nil {
+		return types.Upgrade{}, types.Channel{}, err
+	}
+
+	// if the counterparty sequence is not equal to the current sequence, then either the counterparty chain is out-of-sync or
+	// the message is out-of-sync and we write an error receipt with our own sequence so that the counterparty can update
+	// their sequence as well.
+	// We must then increment our sequence so both sides start the next upgrade with a fresh sequence.
+	if counterpartyUpgradeSequence < channel.UpgradeSequence {
+		errorReceipt := types.NewErrorReceipt(channel.UpgradeSequence, errorsmod.Wrapf(types.ErrInvalidUpgrade, "counterparty chain upgrade sequence <= upgrade sequence (%d <= %d)", counterpartyUpgradeSequence, channel.UpgradeSequence))
+		channel.UpgradeSequence++
+		// TODO: emit error receipt events
+		k.SetUpgradeErrorReceipt(ctx, portID, channelID, errorReceipt)
+
+		// if this code path is reached in a crossing hellos scenario, we must reset the channel state to OPEN
+		if channel.State == types.INITUPGRADE {
+			channel.State = types.OPEN
+		}
+
+		k.SetChannel(ctx, portID, channelID, channel)
+
+		return types.Upgrade{}, types.Channel{}, errorsmod.Wrapf(types.ErrInvalidUpgrade, "counterparty chain upgrade sequence <= upgrade sequence (%d <= %d)", counterpartyUpgradeSequence, channel.UpgradeSequence)
+	}
+
+	var proposedUpgrade types.Upgrade
+
+	// non-crossing hellos case
+	if channel.State == types.OPEN {
+		if counterpartyUpgradeSequence > channel.UpgradeSequence {
+			channel.UpgradeSequence = counterpartyUpgradeSequence - 1
+			k.SetChannel(ctx, portID, channelID, channel)
+		}
+		// ChanUpgradeInit will increment the upgrade sequence
+		proposedUpgrade, err = k.ChanUpgradeInit(
```

Comment:
I think we should have helper functions: `initUpgradeHandshake`, `blockUpgradeHandshake`, and `openHandshake` and compose them together in these handshake functions as shown in the spec PR: https://github.com/cosmos/ibc/pull/967/files

Reply:
maybe we can do that refactor after this PR is in?

Reply:
I'm a fan of splitting it up into helper functions, but am fine with a follow up